enhancements/dns/configurable-dns-pod-placement.md

```diff
@@ -0,0 +1,359 @@
+      node-role.kubernetes.io/infra: ""
+    tolerations:
+    - effect: NoSchedule
+      key: "node-role.kubernetes.io/infra"
+      operator: Exists
+```
+
+Note that the new `NodeSelector` field has type `map[string]string`, rather than
+`*metav1.LabelSelector`, to match the type on the DaemonSet's
+`spec.template.spec.nodeSelector` field's type.  Thus the user can specify label
+key-value pairs to match but cannot specify arbitrary match expressions as using
+`metav1.LabelSelector` would allow.
+
+### Validation
+
+Omitting `spec.nodePlacement` or its subfields specifies the default behavior.
+
+The API validates that `spec.nodePlacement.nodeSelector`, if specified, is a
+valid node selector and that `spec.nodePlacement.tolerations`, if specified, is
+a list of valid tolerations.
+
+### User Stories
+
+#### As a cluster administrator, I must comply with a security policy that prohibits communication among worker nodes
+
+To satisfy this use-case, the cluster administrator can specify a node selector
+that includes only control-plane nodes, using the new
+`spec.nodePlacement.nodeSelector` API field as follows:
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: DNS
+metadata:
+  name: default
+spec:
+  nodePlacement:
+    nodeSelector:
+      kubernetes.io/os: linux
+      node-role.kubernetes.io/master: ""
+```
+
+#### As a cluster administrator, I want to allow DNS Pods to run on nodes that have a taint that has key "dns-only" and effect `NoSchedule`
+
+To satisfy this use-case, the cluster administrator can specify a toleration for
+the taint in question as follows:
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: DNS
+metadata:
+  name: default
+spec:
+  nodePlacement:
+    tolerations:
+    - effect: NoSchedule
+      key: "dns-only"
+      operator: Exists
+```
+
+### Implementation Details
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-dns-operator
+
+The DNS operator is modified to manage both the "dns-default" DaemonSet and the
+"node-resolver" DaemonSet, both in the "openshift-dns" namespace.  The
+"dns-node-resolver" container is removed from the "dns-default" DaemonSet if it
+already exists, as are any tolerations and label selectors that are not
+configured per the new API.  The operator is modified to apply the configured
+tolerations and node label selectors to the "dns-default" DaemonSet.  The
+"dns-node-resolver" is configured to tolerate all taints (as the "dns-default"
+DaemonSet does in OpenShift 4.7 and earlier) and run on all Linux nodes.
+
+### Risks and Mitigations
+
+A cluster administrator could configure a node selector, or taint all nodes, in
+such a way that DNS Pods could not be scheduled to any node, rendering the DNS
+service unavailable.  Because the DNS service is critical to other cluster
+components including OAuth, fixing misconfigured DNS Pod placement parameters
+could be impossible for the cluster administrator to do.
```

Comment:
Is this only scenario that could block fixing this the user being unable to login?  I’m skeptical that a one time check for placement is the right approach, vs (for example) always ensuring dns runs on the control plane nodes if there are no valid worker nodes in use (or separately as a fallback).  Emulating the scheduler is probably not a great idea (vs detecting zero pods at runtime or always running the previous topology until the new topology is in place, or “failing open” if no pods are detected).  Generally doing any sort of static detection of schedulability adds fragility to a system.

Reply:
Yeah, if DNS isn't available to resolve `kubernetes.default.svc.cluster.local`, then presumably that could break OAuth.  If the administrator didn't already have an OAuth token handy, the problem might then be impossible to resolve.  If the user did have an OAuth token, it should be possible to patch the CRD to revert the problematic node-placement parameters.  I agree that trying to infer what the scheduler will do is an imperfect mitigation.  We could add some logic to the reconcile loop to detect if the operator has been reporting degraded for &gt;_n_ minutes and has 0 DNS pods; would it be valid for the operator to revert `spec.nodePlacement`, or does the operator need to maintain some state to track that the current value in `spec.nodePlacement` resulted in having 0 pods and needs to be ignored?